Issue #6552 - Windows Cleaner+FileDisposer for FileBufferedResponseHandler

joakime opened the pull request:
Proposal of using `java.lang.ref.Cleaner` with `FileBufferedResponseHandler`

joakime committed:
Issue #6552 - Example of java.lang.ref.Cleaner usage

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

joakime committed:
Issue #6552 - Some logging changes

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

gregw commented on jetty-util/src/main/java/org/eclipse/jetty/util/PhantomCleaner.java:
```diff
@@ -0,0 +1,71 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public final class PhantomCleaner
```
I know it the JVM class uses "Phantom" as a concept, but I don't like the name here.   `FileGcCleaner` would be better.

sbordet replied:
This is a generic mechanism, not only for files, so `JVMCleaner` or `ReferenceCleaner` or just `Cleaner`.

gregw commented on jetty-server/src/main/java/org/eclipse/jetty/server/handler/FileBufferedResponseHandler.java:
```diff
@@ -225,7 +216,21 @@ protected void onCompleteFailure(Throwable cause)
             };
+            // once icb is GC'd, cleanup the File.
+            if (_filePath != null)
+                PhantomCleaner.register(icb, new PhantomCleaner.FileDispose(_filePath));
```
Do we only want to do delayed deletion on windows? or do we want to use the same mechanism everywhere?

Perhaps we could try to delete here and if that fails then register the cleaner?

gregw replied:
Also is the icb really the right object?  Surely it should be the buffer?   icb doesn't even have a reference to the buffer.

sbordet replied:
@gregw it cannot be the buffer because we may iterate and have multiple buffers. Using the ICB guarantees that when the ICB is done, then all the buffers are also done and we register with the cleaner only once.
It is also critical that we register a reference to an object that is not strongly referenced, and the ICB reference satisfies this condition.

This mechanism will delay the file deletion in all cases, not just Windows.
Trying to delete and if fails use the cleaner will introduce a race as the cleaner may run at any moment... but `Files.deleteIfExists()` should take care of that.

sbordet commented on jetty-server/src/main/java/org/eclipse/jetty/server/handler/FileBufferedResponseHandler.java:
```diff
@@ -225,7 +216,21 @@ protected void onCompleteFailure(Throwable cause)
+                    sb.append(", pos=").append(_pos);
+                    sb.append(", last=").append(_last);
+                    sb.append(']');
+                    return sb.toString();
```
Use `String.format()` and `getClass().getSimpleName()`.

sbordet commented on jetty-util/src/main/java/org/eclipse/jetty/util/PhantomCleaner.java:
```diff
@@ -0,0 +1,71 @@
+            final StringBuilder sb = new StringBuilder("FileDispose[");
+            sb.append("filename='").append(filename).append('\'');
+            sb.append(']');
+            return sb.toString();
```
Please use `String.format()` and `getClass().getSimpleName()`.

sbordet commented on jetty-util/src/main/java/org/eclipse/jetty/util/PhantomCleaner.java:
```diff
@@ -0,0 +1,71 @@
+        CLEANER.register(obj, action);
+    }
+
+    public static class FileDispose implements Runnable
```
This is really a file *delete*, "dispose" is too generic. I would rename to `DeleteFile`.

sbordet commented on jetty-util/src/main/java/org/eclipse/jetty/util/PhantomCleaner.java:
```diff
@@ -0,0 +1,71 @@
+
+    public static void register(Object obj, Runnable action)
+    {
+        LOG.info("register({}, {})", obj, action);
```
`Log.debug()`, not `info()`.

sbordet commented on jetty-util/src/main/java/org/eclipse/jetty/util/PhantomCleaner.java:
```diff
@@ -0,0 +1,71 @@
+        public FileDispose(Path file)
+        {
+            // storing this reference as a String to allow GC to work.
+            this.filename = file.toAbsolutePath().toString();
```
There is no need to store the `Path` as a string, the comment is wrong as this is just the action.

sbordet commented on jetty-util/src/main/java/org/eclipse/jetty/util/PhantomCleaner.java:
```diff
@@ -0,0 +1,71 @@
+            Path file = Paths.get(filename);
+            try
+            {
+                LOG.info("Cleaning {}", file);
```
`LOG.debug()`, not `info()`.

joakime committed:
Issue #6552 - Changes requested during review

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

joakime commented on jetty-server/src/test/java/org/eclipse/jetty/server/handler/FileBufferedResponseHandlerTest.java:
```diff
@@ -119,6 +125,14 @@ public void after() throws Exception
+        System.gc();
+        Thread.sleep(1000);
+        System.gc();
+    }
```
This temporary, and is only used to demonstrate a wart.
It will be removed.

joakime committed:
Issue #6552 - Speculative delete and register

Always register for cleanable, but only if a new
temp file is created.
Expose CleanableFuture for testing reasons

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

sbordet commented on jetty-server/src/test/resources/jetty-logging.properties:
```diff
@@ -4,3 +4,4 @@
 #org.eclipse.jetty.servlet.LEVEL=DEBUG
 #org.eclipse.jetty.server.ConnectionLimit.LEVEL=DEBUG
 #org.eclipse.jetty.server.AcceptRateLimit.LEVEL=DEBUG
+org.eclipse.jetty.util.ReferenceCleaner.LEVEL=DEBUG
```
Comment this out.

sbordet commented on jetty-server/src/main/java/org/eclipse/jetty/server/handler/FileBufferedResponseHandler.java:
```diff
@@ -225,6 +220,14 @@ protected void onCompleteFailure(Throwable cause)
+                @Override
+                public String toString()
+                {
+                    return String.format("FileBufferedInterceptor@%X[fileLength=%d,pos=%d,last=%b]",
```
Don't hardcode the class name, use `getClass().getSimpleName()`.
Also, the hashCode should be `%x`, not `%X`.

joakime replied:
This is an anonymous class.

joakime replied:
I'll just remove this one, as we moved the register from this anonymous class to the outer interceptor anyway.

sbordet commented on jetty-server/src/main/java/org/eclipse/jetty/server/handler/FileBufferedResponseHandler.java:
```diff
@@ -225,6 +220,14 @@ protected void onCompleteFailure(Throwable cause)
+                public String toString()
+                {
+                    return String.format("FileBufferedInterceptor@%X[fileLength=%d,pos=%d,last=%b]",
+                        this.hashCode(), fileLength, _pos, _last
```
Drop `this.`, unnecessary.
Also, field `fileLength` should be prefixed with `_` for consistency.

sbordet commented on jetty-util/src/main/java/org/eclipse/jetty/util/ReferenceCleaner.java:
```diff
@@ -0,0 +1,92 @@
+        @Override
+        public String toString()
+        {
+            return String.format("%s@%x[file='%s']", this.getClass().getSimpleName(), this.hashCode(), file);
```
Drop `this.`, unnecessary.

joakime committed:
Issue #6552 - Changes from review

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

joakime committed:
Issue #6552 - Removing toString

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

sbordet reviewed (approved):
LGTM.
However, if we decide to not use mapping for Windows, then this PR is moot as it is not necessary for Linux.

lachlan-roberts commented on jetty-server/src/main/java/org/eclipse/jetty/server/handler/FileBufferedResponseHandler.java:
```diff
@@ -106,18 +109,9 @@ protected void dispose()
-                }
-                _filePath = null;
-            }
+            ReferenceCleaner.DeleteFile.delete(_filePath);
```
Maybe we should only register the file with the Cleaner if this delete fails. This way we should not have anything registered with the Cleaner except on windows.

joakime replied:
I tried that, the .reset() and .resetBuffer() use cases made that approach difficult.
Besides, the line you flagged just deletes the file directly, no register step.

lachlan-roberts replied:
> Besides, the line you flagged just deletes the file directly, no register step.

I meant to only register here if this line fails to delete the file, and to not register it with the Cleaner when the file is created.


> I tried that, the .reset() and .resetBuffer() use cases made that approach difficult.

What do you mean by the .reset() and .resetBuffer() cases?

gregw commented:
@joakime can you resolve conflicts and get a clean CI build

joakime commented:
I'll need to merge/rebase this PR.
